### PR TITLE
add migration guide

### PR DESCRIPTION
**Pull request by sokra**

### CODE REVIEW

File: src/content/guides/migrating.md [hunk 1]
```diff
@@ -16,12 +16,117 @@ contributors:
...
 
-W> This content may be moved to the blog post in the near future as version 2 has been out for a while. On top of that, version 3 was recently released and version 4 is on the horizon. As noted above, folks should instead to refer to [the changelog](https://github.com/webpack/webpack/releases) for migrations.
+``` diff
++mode: "production",
+```
+
+``` diff
++mode: "development",
+```
+
```
Comment (nveenjain) [inline src/content/guides/migrating.md]: Also, we can pass mode: none, right?

**Comment by montogeek** [comment]
@EugeneHlushko You may want to add something here?

File: src/content/guides/migrating.md [hunk 2]
```diff
@@ -16,12 +16,117 @@ contributors:
...
+``` diff
+-{
+-  test: /\.json$/,
+-  loader: "json-loader"
+-}
+```
+
+## module.loaders
+
+`module.loaders` was deprecated since webpack 2 and is now removed in favor of `module.rules`.
```
Comment (EugeneHlushko) [inline src/content/guides/migrating.md]: were deprecated

File: src/content/guides/migrating.md [hunk 3]
```diff
@@ -16,12 +16,117 @@ contributors:
...
+
+When generating the HTML from the stats you can use `optimization.splitChunks.chunks: "all"` which is the optimal configuration in most cases.
+
+## import() and CommonJS
+
+When using `import()` to load non-ESM the result has changed in webpack 4. You now need access the `default` property to get the value of `module.exports`.
+
+## json and loaders
+
+When using a custom loader to transform `.json` files you now need to change the module type:
```
Comment (EugeneHlushko) [inline src/content/guides/migrating.md]: `type`

File: src/content/guides/migrating.md [hunk 4]
```diff
@@ -16,12 +16,117 @@ contributors:
...
+
+The `CommonsChunkPlugin` was removed. Instead the `optimization.splitChunks` options can be used.
+
+See documentation for `optimization.splitChunks` to details. The default configuration may already suit your needs.
+
+When generating the HTML from the stats you can use `optimization.splitChunks.chunks: "all"` which is the optimal configuration in most cases.
+
+## import() and CommonJS
+
+When using `import()` to load non-ESM the result has changed in webpack 4. You now need access the `default` property to get the value of `module.exports`.
```
Comment (EugeneHlushko) [inline src/content/guides/migrating.md]: Now you need to access the `default` property to get the value of `module.exports`.
__non-esm.js__
```javascript
module.exports = {
  sayHello: () => { console.log('hello world') }
};
```

__example.js__
```javascript
function sayHello() {
  import('./non-esm.js').default.sayHello();
}
sayHello();
```

File: src/content/guides/migrating.md [hunk 5]
```diff
@@ -16,12 +16,117 @@ contributors:
...
+These plugins were deprecated and are now removed:
+
+``` diff
+-new NoErrorsPlugin(),
+-new NewWatchingPlugin(),
+```
+
+## CommonsChunkPlugin
+
+The `CommonsChunkPlugin` was removed. Instead the `optimization.splitChunks` options can be used.
```
Comment (EugeneHlushko) [inline src/content/guides/migrating.md]: [`optimization.splitChunks`](/configuration/optimization/#optimization-splitchunks)

Comment (EugeneHlushko) [reply src/content/guides/migrating.md]: its getting auto formatted, but i suggest adding links to the docs.

File: src/content/guides/migrating.md [hunk 6]
```diff
@@ -16,12 +16,117 @@ contributors:
...
-The following sections describe the major changes from webpack 1 to 2.
+This guide only shows major changes that affect end users. For more details please see [the changelog](https://github.com/webpack/webpack/releases).
+
+# Migration from 3 to 4
+
+[Changelog](https://github.com/webpack/webpack/releases/tag/v4.0.0)
+
+## node.js 4
+
+If you are still using node.js 4 or lower, you need to upgrade your node.js installation
```
Comment (EugeneHlushko) [inline src/content/guides/migrating.md]: Node.js

File: src/content/migrate/3.md [hunk 7]
```diff
@@ -1,6 +1,6 @@
 ---
-title: To webpack v2, v3
-sort: 1
+title: To v2, v3
```
Comment (montogeek) [inline src/content/migrate/3.md]: I don't understand previous title, neither this one. It is too cryptic :D

File: src/content/migrate/4.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
+---
+title: To v4 from v3
```
Comment (montogeek) [inline src/content/migrate/4.md]: We should keep titles consistent.

File: src/content/migrate/3.md (hunk 7 above)
Comment (EugeneHlushko) [reply src/content/migrate/3.md]: when you open the page it makes sense: https://webpack.js.org/migrate/

**Comment by EugeneHlushko** [comment]
@montogeek updated

**Comment by montogeek** [comment]
Thanks!

**Pull request merged**